source/iterator/iterator.go

```diff
@@ -0,0 +1,311 @@
+	"github.com/conduitio-labs/conduit-connector-redshift/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/multierr"
+)
+
+const (
+	// metadataFieldTable is a name of a record metadata field that stores a Redshift table name.
+	metadataFieldTable = "redshift.table"
+	// keySearchPath is a key of get parameter of a datatable's schema name.
+	keySearchPath = "search_path"
+)
+
+// Iterator is an implementation of an iterator for Amazon Redshift.
+type Iterator struct {
+	db       *sqlx.DB
+	rows     *sqlx.Rows
+	position *Position
+
+	// table is a table name.
+	table string
+	// keyColumns is a name of the column that iterator will use for setting key in record.
+	keyColumns []string
+	// orderingColumn is a name of the column that iterator will use for sorting data.
+	orderingColumn string
+	// batchSize is a size of rows batch.
+	batchSize int
+	// columnTypes is a map with all column types.
+	columnTypes map[string]string
+}
+
+// New creates a new instance of the iterator.
+func New(ctx context.Context, driverName string, pos *Position, config config.Source) (*Iterator, error) {
+	var err error
+
+	iterator := &Iterator{
+		position:       pos,
+		table:          config.Table,
+		keyColumns:     config.KeyColumns,
+		orderingColumn: config.OrderingColumn,
+		batchSize:      config.BatchSize,
+	}
+
+	iterator.db, err = sqlx.Open(driverName, config.DSN)
+	if err != nil {
+		return nil, fmt.Errorf("open db connection: %w", err)
+	}
+
+	if iterator.position.LastProcessedValue == nil {
+		latestSnapshotValue, latestValErr := iterator.latestSnapshotValue(ctx)
+		if latestValErr != nil {
+			return nil, fmt.Errorf("get latest snapshot value: %w", latestValErr)
+		}
+
+		if config.Snapshot {
+			iterator.position.LatestSnapshotValue = latestSnapshotValue
+		} else {
+			iterator.position.LastProcessedValue = latestSnapshotValue
+		}
+	}
+
+	u, err := url.Parse(config.DSN)
+	if err != nil {
+		return nil, fmt.Errorf("parse dsn: %w", err)
+	}
+
+	schema := u.Query().Get(keySearchPath)
+
+	err = iterator.populateKeyColumns(ctx, schema)
+	if err != nil {
+		return nil, fmt.Errorf("populate key columns: %w", err)
+	}
+
+	iterator.columnTypes, err = columntypes.GetColumnTypes(ctx, iterator.db, iterator.table, schema)
+	if err != nil {
+		return nil, fmt.Errorf("get column types: %w", err)
+	}
+
+	err = iterator.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	return iterator, nil
+}
+
+// HasNext returns a bool indicating whether the source has the next record to return or not.
+func (iter *Iterator) HasNext(ctx context.Context) (bool, error) {
+	if iter.rows != nil && iter.rows.Next() {
+		return true, nil
+	}
+
+	if err := iter.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	if iter.rows.Next() {
+		return true, nil
+	}
+
+	if iter.position.LatestSnapshotValue != nil {
```

Comment:
I misunderstood this part. Could you please tell me what we do here, we can discuss it on next  meeting

Reply:
I've added comments.